neuraxle/steps/flow.py

```diff
@@ -0,0 +1,199 @@
+        :return: step, data_container
+        :type: DataContainer
+        """
+        self._nullify_hyperparams()
+        return data_container
+
+    def _nullify_hyperparams(self):
+        """
+        Nullify wrapped step hyperparams using hyperparams_space.nullify().
+        """
+        hyperparams_space = self.wrapped.get_hyperparams_space()
+        self.wrapped.set_hyperparams(hyperparams_space.nullify())
+
+
+CHOOSE_ONE_OR_MANY_STEPS_OF_CHOICE_HYPERPARAM = 'choice'
+
+
+class ChooseOneOrManyStepsOf(Pipeline):
+    """
+    A pipeline to allow choosing many steps using an hyperparameter.
+
+    Example usage :
+
+    .. code-block:: python
+
+        p = Pipeline([
+            ChooseOneOrManyStepsOf([
+                ('a', Identity()),
+                ('b', Identity())
+            ])
+        ])
+        p.set_hyperparams({
+            'ChooseOneOrManyStepsOf': {
+                'choice': {
+                    'a': True
+                    'b': False
+                }
+            }
+        })
+
+    .. seealso::
+        :class:`Pipeline`
+        :class:`Nullify`
+    """
+
+    def __init__(self, steps, hyperparams=None):
+        Pipeline.__init__(self, steps)
+
+        if hyperparams is None:
+            hyperparams = HyperparameterSamples({
+                'choice': {}
+            })
+
+        self.set_hyperparams(hyperparams)
+
+    def set_hyperparams(self, hyperparams: Union[HyperparameterSamples, OrderedDict, dict]) -> BaseStep:
+        """
+        Set hyperparams for step selection, and nullify the steps that are not chosen.
+
+        :param hyperparams: hyperparams
+        :type hyperparams: HyperparameterSamples
+        :return: self
+        :rtype: BaseStep
+        """
+        super().set_hyperparams(hyperparams=hyperparams)
+
+        if CHOOSE_ONE_OR_MANY_STEPS_OF_CHOICE_HYPERPARAM not in self.hyperparams:
+            raise ValueError('\'choice\' hyperparam not set in {0} hyperparams'.format(self.name))
+
+        for key in self.hyperparams[CHOOSE_ONE_OR_MANY_STEPS_OF_CHOICE_HYPERPARAM].keys():
+            if key not in self.keys():
+                raise ValueError('Invalid Choosen Step {0} in {1}'.format(key, self.name))
+
+        for key in self.keys():
+            if key not in self.hyperparams[CHOOSE_ONE_OR_MANY_STEPS_OF_CHOICE_HYPERPARAM].keys():
+                self.hyperparams[CHOOSE_ONE_OR_MANY_STEPS_OF_CHOICE_HYPERPARAM][key] = False
+                warning.warn('You have not set any choice hyperparameter for {0} in {1}. Setting choice to False by default.'.format(key, self.name))
+
+        self.nullify_steps_that_are_not_chosen()
+
+        return self
+
+    def nullify_steps_that_are_not_chosen(self):
+        """
+        Nullify steps that are not chosen using :class:`Nullify` step wrapper.
+        """
+        for step_name, is_chosen in self.hyperparams[CHOOSE_ONE_OR_MANY_STEPS_OF_CHOICE_HYPERPARAM].items():
+            if not is_chosen:
```

Comment:
There is a OOP problem here. This is not the responsibility of this class. Solution: 

1. Delete `nullify_steps_that_are_not_chosen`
2. Rename `class Nullify` to `class Optional`
3. Make `ChooseOneOrManyStepsOf` dynamically pre-wrap all the provided steps at construction with an Optional. 

This makes the set_hyperparams of the Optional change the behavior of the optional class itself. 

As a consequence, the `Boolean()` Hyperparameter distribution will be used one time in each Optional, and if set to true, it will make the wrapped things nullified.

Reply:
I don't even like my solution. Because the `class ChooseOneStepOf(TruncableSteps):` will need to force to choose exactly just one on an `rvs` and the Optional wrapper won't work. 

Well, still, I really like the optional and I maybe prefer it more than a Nullified. What's cool with optionals is that it becomes way easier to choose one or many steps in completely different contexts.

Reply:
After re-thinking, I like the optional. Perhaps the ChooseOneStepOf could work completely differently in the worst case, not using optionals at all. The `ChooseOneStepOf` could be even called `StepChoice` so much it's different.